applications/minotari_node/src/stratum/block_template_repository.rs

```diff
@@ -0,0 +1,329 @@
+    pub target: u64,
+    pub seed_hash: Option<Vec<u8>>,
+    pub prev_block_hash: Vec<u8>,
+}
+#[async_trait::async_trait]
+pub trait BlockTemplateRepository {
+    async fn get_block_template(
+        &self,
+        algo: PowAlgorithm,
+        solo_address: Option<TariAddress>,
+    ) -> Result<BlockTemplate, anyhow::Error>;
+    async fn get_tip(&self) -> Result<(u64, Vec<u8>), anyhow::Error>;
+}
+
+#[derive(Clone)]
+pub struct DefaultBlockTemplateRepository {
+    node_service: LocalNodeCommsInterface,
+    consensus_rules: BaseConsensusManager,
+    block_templates: Arc<DashMap<Vec<u8>, Block>>,
+}
+
+impl DefaultBlockTemplateRepository {
+    pub fn new(node_service: LocalNodeCommsInterface, consensus_rules: BaseConsensusManager) -> Self {
+        Self {
+            node_service,
+            consensus_rules,
+            block_templates: DashMap::new().into(),
+        }
+    }
+
+    pub async fn start(
+        &self,
+        mut shutdown_signal: ShutdownSignal,
+        mut job_receiver: SubmitJobQueueReceiver,
+    ) -> Result<(), anyhow::Error> {
+        let mut node_service = self.node_service.clone();
+        let templates = self.block_templates.clone();
+        tokio::spawn(async move {
+            loop {
+                select! {
+                    _ = shutdown_signal.wait() => {
+                        debug!(target: LOG_TARGET, "Shutting down Block Template Repository");
+                        break;
+                    }
+                    Some((job, responder)) = job_receiver.recv() => {
+                        debug!(target: LOG_TARGET, "Received job submission for job ID: {}", job.job_id);
+                        let block_template =templates.get(&job.original_mining_hash);
+                        if let Some(block) = block_template {
+                            debug!(target: LOG_TARGET, "Found block template for job ID: {}", job.job_id);
+
+                            let mut block = block.clone();
+                            info!(target: LOG_TARGET, "Submitting block at height: {} for job ID: {} algo: {}", block.header.height, job.job_id, job.pow_algo);
+                            match job.pow_algo {
+                              PowAlgorithm::RandomXT => {
+                                block.header.nonce = job.nonce;
+                              }
+                              PowAlgorithm::Sha3x => {
+                                block.header.nonce = job.nonce.to_be();
+                              }
+                              PowAlgorithm::Cuckaroo => {
+                                block.header.nonce = job.nonce;
+                                block.header.pow.pow_algo = PowAlgorithm::Cuckaroo;
+                                info!(target: LOG_TARGET, "Packing nonces for Cuckaroo job ID: {} with nonces: {:?}", job.job_id, job.cuckaroo_nonces);
+                                block.header.pow.pow_data = match pack_nonces(&job.cuckaroo_nonces, 29).try_into() {
```

Comment:
We can use the consensus constants here to replace `29`
```suggestion
                                block.header.pow.pow_data = match pack_nonces(&job.cuckaroo_nonces, 29).try_into() {
```